src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs

```diff
@@ -601,46 +597,125 @@ LIMIT 1
+    WHERE from_state_version <= @stateVersion AND global_entity_id = @entityId AND resource_entity_id = vah.fungible_resource_entity_id AND vault_entity_id = vah.vault_entity_id
     ORDER BY from_state_version DESC
     LIMIT 1
-) evh ON TRUE
-INNER JOIN entities e ON p.non_fungible_resource_entity_id = e.id
-GROUP BY ResourceEntityGlobalAddress, TotalCount;
+) vh ON TRUE
+INNER JOIN entities er ON vah.fungible_resource_entity_id = er.id
+INNER JOIN entities ev ON vah.vault_entity_id = ev.id;
+",
+            parameters: new
+            {
+                stateVersion = ledgerState.StateVersion,
+                entityId = entityId,
+                resourceLimit = resourceLimit,
+                vaultLimit = vaultLimit,
+            },
+            cancellationToken: token);
+
+        long? resourcesTotalCount = 0;
+
+        var resources = new Dictionary<GlobalAddress, GatewayModel.FungibleResourcesCollectionItemVaultAggregated>();
+
+        foreach (var vm in await _dbContext.Database.GetDbConnection().QueryAsync<FungibleVaultViewModel>(cd))
+        {
+            resourcesTotalCount = vm.ResourceTotalCount;
+
+            if (!resources.TryGetValue(vm.ResourceEntityGlobalAddress, out var existingRecord))
+            {
+                var vaultNextCursor = vm.VaultTotalCount > vaultLimit
+                    ? new GatewayModel.EntityFungiblesCursor(vaultLimit).ToCursorString()
+                    : null;
+
+                existingRecord = new GatewayModel.FungibleResourcesCollectionItemVaultAggregated(
+                    resourceAddress: vm.ResourceEntityGlobalAddress,
+                    vaults: new GatewayModel.FungibleResourcesCollectionItemVaultAggregatedVault(
+                        totalCount: vm.VaultTotalCount,
+                        nextCursor: vaultNextCursor,
+                        items: new List<GatewayModel.FungibleResourcesCollectionItemVaultAggregatedVaultItem>()));
+
+                resources[vm.ResourceEntityGlobalAddress] = existingRecord;
+            }
+
+            existingRecord.Vaults.Items.Add(new GatewayModel.FungibleResourcesCollectionItemVaultAggregatedVaultItem(
+                vaultAddress: vm.VaultAddress,
+                amount: TokenAmount.FromSubUnitsString(vm.Balance).ToString()));
+        }
+
+        var nextCursor = resourcesTotalCount > resourceLimit
+            ? new GatewayModel.EntityFungiblesCursor(resourceLimit).ToCursorString()
+            : null;
+
+        return new GatewayModel.FungibleResourcesCollection(resourcesTotalCount, null, nextCursor, resources.Values.Cast<GatewayModel.FungibleResourcesCollectionItem>().ToList());
+    }
+
+    /// <summary>
+    /// Returns a paginable resource collection with total (aggregated) count per resource.
+    /// </summary>
+    private async Task<GatewayModel.NonFungibleResourcesCollection> GetNonFungiblesSlice(long entityId, int offset, int limit, GatewayModel.LedgerState ledgerState, CancellationToken token)
```

Comment:
can we rename it to `GetAggregatedNonFungibleSlice` so we could easily distinguish it from `GetNonFungiblesSlicePerVault`?

Reply:
Renamed to `*PerResource` to match `*PerVault` scheme.